README.md
# Disqus

The Disqus Gem helps you easily integrate the [Disqus](http://disqus.com)
commenting system into your Ruby-based website. It includes a Ruby wrapper for
[Disqus's REST API](http://wiki.disqus.net/API), and views helpers to generate
their Javasript widgets.

## What is Disqus?

From the Disqus website:

> "Disqus, pronounced "discuss", is a service and tool for web comments and
> discussions. The Disqus comment system can be plugged into any website, blog,
> or application. Disqus makes commenting easier and more interactive, while
> connecting websites and commenters across a thriving discussion community."
> 
> "Disqus is a free service to the general public with absolutely no inline advertisements."


### Getting it

    gem install disqus
    
### Using it

#### Configure it

  Disqus::defaults[:account] = "my_disqus_account"
  # Optional, only if you're using the API
  Disqus::defaults[:api_key] = "my_disqus_api_key"


#### Options

    :api_key         => "" # your api key
    :account         => "", # your disqus account
    :developer       => false, # allows threads to work on localhost
    :container_id    => 'disqus_thread', # desired thread container
    :avatar_size     => 48, # squared pixel size of avatars
    :color           => "grey", # theme color
    :default_tab     => "popular", # default widget tab
    :hide_avatars    => false, # hide or show avatars
    :hide_mods       => true, # hide or show moderation
    :num_items       => 15, # number of comments to display
    :show_powered_by => true, # show or hide powered by line
    :orientation     => "horizontal" # comment orientation

#### Show the comment threads widget on a post page

    # Loads the commenting system
    disqus_thread
  
    # Sets the inner html to the comment count for any links on the page that
    # have the anchor "disqus_thread". For example, "View Comments" below would
    # be replaced by "1 comment" or "23 comments" etc.
    # <a href="http://my.website/article-permalink#disqus_thread">View Comments</a>
    # <a href="http://my.website/different-permalink#disqus_thread">View Comments</a>
    disqus_comment_counts
  
#### Show the combo widget on a post page

    disqus_combo(:color => "blue", :hide_mods => false, :num_items => 20)
  
#### Show the comment count on a permalink

    link_to("Permalink", post_path(@post, :anchor => "disqus_thread"))
    ...
    disqus_comment_counts
  
#### Work with the Disqus API

See the Disqus::Api class for more info on the Disqus API. You can also read the
[Disqus developer info here](http://disqus.com/docs/api/).

### Hack it

Github repository: 
  
[http://github.com/norman/disqus](http://github.com/norman/disqus)

### Learn more about Disqus:

[http://disqus.com](http://disqus.com)

### Thanks to the following contributors:

* {Matt Van Horn}[http://github.com/mattvanhorn] - Disqus API
* {Quin Hoxie}[http://github.com/qhoxie] - Merb support

### Legal Stuff

The Disqus Ruby gem was not created by, nor is officially supported by
Disqus.com or Big Head Labs, Inc. Use it at your own risk and your own
responsibility under the terms of the MIT License.

Copyright (c) 2008-2010 Norman Clarke. Released under the MIT license

CHANGELOG
Author: Norman Clarke <norman@randomba.org>
Date:   Tue Aug 26 17:11:32 2008 -0300

    Removed entities, not sure how they get formatted. Bleh.

Author: Norman Clarke <norman@randomba.org>
Date:   Tue Aug 26 17:09:31 2008 -0300

    Tweak entities in README and add ROOT_PATH to js.

Author: Norman Clarke <norman@randomba.org>
Date:   Tue Aug 26 17:05:52 2008 -0300

    Initial commit.


lib/disqus/widget.rb
module Disqus

  class Widget
  
    class Error < StandardError ; end
    
    VALID_COLORS = ['blue', 'grey', 'green', 'red', 'orange']
    VALID_NUM_ITEMS = 5..20
    VALID_DEFAULT_TABS = ['people', 'recent', 'popular']
    VALID_AVATAR_SIZES = [24, 32, 48, 92, 128]
    VALID_ORIENTATIONS = ['horizontal', 'vertical']

    ROOT_PATH = 'http://disqus.com/forums/%s/'
    THREAD = ROOT_PATH + 'embed.js'
    COMBO = ROOT_PATH + 'combination_widget.js?num_items=%d&color=%s&default_tab=%s'
    RECENT = ROOT_PATH + 'recent_comments_widget.js?num_items=%d&avatar_size=%d'
    POPULAR = ROOT_PATH + 'popular_threads_widget.js?num_items=%d'
    TOP = ROOT_PATH + 'top_commenters_widget.js?num_items=%d&avatar_size=%d&orientation=%s'
    class << self
      
      # Show the main Disqus thread widget. Options:
      # * <tt>account:</tt> Your Discus account (required).
      def thread(opts = {})
        opts = Disqus::defaults.merge(opts)
        opts[:view_thread_text] ||= "View the discussion thread"
        validate_opts!(opts)
        s = '<div id="disqus_thread"></div>'
        s << '<script type="text/javascript" src="' + THREAD + '"></script>'
        s << '<noscript><a href="http://%s.disqus.com/?url=ref">'
        s << opts[:view_thread_text]
        s << '</a></noscript>'
        if opts[:show_powered_by]
          s << '<a href="http://disqus.com" class="dsq-brlink">blog comments '
          s << 'powered by <span class="logo-disqus">Disqus</span></a>'
        end
        s % [opts[:account], opts[:account]]
      end
      
      # Loads Javascript to show the number of comments for the page. Options:
      # * <tt>account:</tt> Your Discus account (required).
      def comment_counts(opts = {})
        opts = Disqus::defaults.merge(opts)        
        validate_opts!(opts)
        s = <<-WHIMPER
        <script type="text/javascript">
        //<[CDATA[
        (function() {
        		var links = document.getElementsByTagName('a');
        		var query = '?';
        		for(var i = 0; i < links.length; i++) {
        			if(links[i].href.indexOf('#disqus_thread') >= 0) {
        				query += 'url' + i + '=' + encodeURIComponent(links[i].href) + '&';
        			}
        		}
        		document.write('<script type="text/javascript" src="#{ROOT_PATH}get_num_replies.js' + query + '"></' + 'script>');
        	})();
        //]]>
        </script>
        WHIMPER
        s % opts[:account]
      end
      
      # Show the main Disqus thread widget. Options:
      # * <tt>account:</tt> Your Discus account (required).
      # * <tt>header:</tt> HTML snipper with header (default h2) tag and text.
      # * <tt>show_powered_by:</tt> Show or hide the powered by Disqus text.
      # * <tt>num_items:</tt>: How many items to show.
      # * <tt>hide_mods:</tt> Don't show moderators.
      # * <tt>hide_avatars:</tt> Don't show avatars.
      # * <tt>avatar_size:</tt> Avatar size.
      def top_commenters(opts = {})
        opts = Disqus::defaults.merge(opts)
        opts[:header] ||= '<h2 class="dsq-widget-title">Top Commenters</h2>'
        validate_opts!(opts)        
        s = '<div id="dsq-topcommenters" class="dsq-widget">'
        s << opts[:header]
        s << '<script type="text/javascript" src="'
        s << TOP
        s << '&hide_avatars=1' if opts[:hide_avatars]
        s << '&hide_mods=1' if opts[:hide_mods]
        s << '"></script>'
        s << '</div>'
        if opts[:show_powered_by]
          s << '<a href="http://disqus.com">Powered by Disqus</a>'
        end
        s % [opts[:account], opts[:num_items], opts[:avatar_size], opts[:orientation]]
      end
      
      # Show the main Disqus thread widget. Options:
      # * <tt>account:</tt> Your Discus account (required).
      # * <tt>header:</tt> HTML snipper with header (default h2) tag and text.
      # * <tt>num_items:</tt>: How many items to show.
      # * <tt>hide_mods:</tt> Don't show moderators.
      def popular_threads(opts = {})
        opts = Disqus::defaults.merge(opts)
        opts[:header] ||= '<h2 class="dsq-widget-title">Popular Threads</h2>'
        validate_opts!(opts)
        s = '<div id="dsq-popthreads" class="dsq-widget">'
        s << opts[:header]
        s << '<script type="text/javascript" src="'
        s << POPULAR
        s << '&hide_mods=1' if opts[:hide_mods]
        s << '"></script>'
        s << '</div>'
        s << '<a href="http://disqus.com">Powered by Disqus</a>' if opts[:show_powered_by]
        s % [opts[:account], opts[:num_items]]
      end
    
      # Show the main Disqus thread widget. Options:
      # * <tt>account:</tt> Your Discus account (required).
      # * <tt>header:</tt> HTML snipper with header (default h2) tag and text.
      # * <tt>num_items:</tt>: How many items to show.
      # * <tt>hide_avatars:</tt> Don't show avatars.
      # * <tt>avatar_size:</tt> Avatar size.
      def recent_comments(opts = {})
        opts = Disqus::defaults.merge(opts)
        opts[:header] ||= '<h2 class="dsq-widget-title">Recent Comments</h2>'
        validate_opts!(opts)
        s = '<div id="dsq-recentcomments" class="dsq-widget">'
        s << opts[:header]
        s << '<script type="text/javascript" src="'
        s << RECENT 
        s << '&hide_avatars=1' if opts[:hide_avatars]
        s << '"></script>'
        s << '</div>'
        if opts[:show_powered_by]
          s << '<a href="http://disqus.com">Powered by Disqus</a>'
        end
        s % [opts[:account], opts[:num_items], opts[:avatar_size]]
      end
    
      # Show the main Disqus thread widget. Options:
      # * <tt>account:</tt> Your Discus account (required).
      # * <tt>num_items:</tt>: How many items to show.
      # * <tt>hide_mods:</tt> Don't show moderators.
      def combo(opts = {})
        opts = Disqus::defaults.merge(opts)
        validate_opts!(opts)
        s = '<script type="text/javascript" src="'
        s << COMBO
        s << '&hide_mods=1' if opts[:hide_mods]
        s << '"></script>' 
        s % [opts[:account], opts[:num_items], opts[:color], opts[:default_tab]]
      end
      
      private

      def validate_opts!(opts)
        raise Error.new("You must specify an :account") if !opts[:account]
        raise Error.new("Invalid color") if opts[:color] && !VALID_COLORS.include?(opts[:color])
        raise Error.new("Invalid num_items") if opts[:num_items] && !VALID_NUM_ITEMS.include?(opts[:num_items])
        raise Error.new("Invalid default_tab") if opts[:default_tab] && !VALID_DEFAULT_TABS.include?(opts[:default_tab])
        raise Error.new("Invalid avatar size") if opts[:avatar_size] && !VALID_AVATAR_SIZES.include?(opts[:avatar_size])
        raise Error.new("Invalid orientation") if opts[:orientation] && !VALID_ORIENTATIONS.include?(opts[:orientation])
      end
      
    end
  
  end

end